### PR TITLE
add compiler hints to reduce number of slice bounds checks

### PR DESCRIPTION
**Pull request by NyaaaWhatsUpDoc**
Just some quick changes that hint to the compiler expected lengths in places to reduce the number of slice bounds checks it needs to do. You can confirm before and after results by using `-gcflags="-d=ssa/check_bce/debug=1"`, or more usefully i was specifically doing `go test -gcflags="-d=ssa/check_bce/debug=1" -v 2>&1 1>/dev/null | grep $fileNameToCheck`

### CODE REVIEW

**Comment by NyaaaWhatsUpDoc** [comment]
in particular a good example of another use of the underscored slice value accessing is in `binary.{LittleEndian,BigEndian}`

File: stmt.go [hunk 1]
```diff
@@ -639,6 +640,14 @@ func (s *Stmt) Columns(dest []any) error {
 	}
 
 	types := util.View(s.c.mod, typePtr, count)
+
+	// hint to the compiler that
+	// it can omit bounds check
+	// accessing types[i] below.
+	if len(types) != len(dest) {
+		return errors.New("incorrect destination count")
```
Comment (ncruces) [inline stmt.go:648]: If this can never happen (I don't think it can):
```go
panic(util.AssertErr())
```

This increases the usefulness of code coverage for me (as my tooling ignores these lines).

**Review by ncruces — APPROVED** [review]
This looks like a great idea, and you covered some of the hot points.

Was it profile guided, should I look elsewhere?

Thanks!

**Comment by NyaaaWhatsUpDoc** [comment]
> This looks like a great idea, and you covered some of the hot points.
> 
> Was it profile guided, should I look elsewhere?
> 
> Thanks!

i wish i could say it was anything as intelligent as having been directed by profile guided optimizations... it was more just i was bored, fiddling around, and noticed an opportunity for improvement :')

**Review by ncruces — APPROVED** [review]

**Comment by ncruces** [comment]
The `func.go` ones can be really hot points, maybe not on “normal” usage (where callbacks are mostly used to read data from disk), but certainly if used in (e.g.) an aggregate function or collation sequence (where a callback may be called millions of times to sort a row set, or calculate statistics).

`Columns` is not _that_ hot, but it's still almost single handedly responsible for this driver being close to as fast in queries as `mattn`.

**Comment by ncruces** [comment]
This lead to a 7% improvement on this [benchmark](https://github.com/ncruces/go-sqlite3/blob/170e1dbebd751d84b6f359d9a188218fac4a163a/ext/stats/stats_test.go#L211C6-L211C24).
Still lots of overhead, but every bit helps, so thanks!
